### PR TITLE
Apply black formatting to the codebase and add enforced status checks on PRs before merging

### LINKED ISSUE #240: Add GPU-enabled CI using resources on the IRIS cluster
There have been recent developments on the IRIS cluster that have enabled the ability to deploy GPU-enabled runners on the IRIS cluster via Github Actions. It would be great to make use of this to get CI for httomo that is more complete, in the sense of being able to run httomo tests with CI that require GPUs.

### LINKED ISSUE #272: Add pull request template
With a lack of CI for the moment, it's easy for developers to forget to run all tests before merging a PR.

One thing that could be helpful is having a checklist of some sort for each PR that must be complete before a merge is possible. One item on the checklist could state that all tests must pass before the PR can be considered for merging.

More info about pull request templates can be found at https://docs.github.com/en/communities/using-templates-to-encourage-useful-issues-and-pull-requests/creating-a-pull-request-template-for-your-repository

### PR DESCRIPTION
**Pull request by dkazanc**
Fixes #240
Fixes #272

Main changes:
- apply black formatting to entire codebase
- update github action workflows
- add required status checks (in repo settings)

## Checklist

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes
- [x] I have made corresponding changes to the documentation

### CODE REVIEW

File: .github/workflows/dev_build_test.yml [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+      env:
+        NVIDIA_VISIBLE_DEVICES: ${{ env.NVIDIA_VISIBLE_DEVICES }}
+
+    defaults:
+      run:
+        shell: bash -l {0}
+
+    steps:
+      - name: Checkout repository code
+        uses: actions/checkout@v3
```
Comment (namannimmo10) [inline .github/workflows/dev_build_test.yml:26]:
```suggestion
        uses: actions/checkout@v4
```
Not so important but, you can fix the warning [here](https://github.com/DiamondLightSource/httomo/actions/runs/9894014080) with this change.

Comment (dkazanc) [reply .github/workflows/dev_build_test.yml]: cheers, will fix that in other files as well.

**Comment by team-gpu** [comment]
Do you perhaps want to add coverage reporting to the CI/CD? It's fairly straightforward with Pytest and you could publish to a platform like [coveralls](https://coveralls.io/). This would add the code coverage stats to each PR, and you can see if a new PR increased or decreased coverage, and the lines hit/missed by the tests. Like [here for example](https://coveralls.io/github/auto-differentiation/xad?branch=main) (C++).

You can also set targets and block the merge until a certain coverage target it hit (or when there is a coverage decrease compared to `main` or something like that).

**Comment by dkazanc** [comment]
@namannimmo10 would you be interested to set the coverage up with a badge as @team-gpu suggests? Now we've got the tests running on iris, we can do that. Will be useful to maintain the coverage levels... Thanks!

**Comment by namannimmo10** [comment]
Sure, will add coverage stats.

**Comment by yousefmoazzam** [comment]
Question for other people's opinion relating to status checks: as mentioned in #273, what do people think about requiring an approved review before merging?

File: .github/pull_request_template.md [hunk 2]
```diff
@@ -0,0 +1,5 @@
+## Checklist
```
Comment (yousefmoazzam) [inline .github/pull_request_template.md:1]: Just a small suggestion, to add a placeholder reference to an issue that the PR is addressing:
```suggestion
Fixes #ISSUE

## Checklist
```

**Comment by team-gpu** [comment]
> Question for other people's opinion relating to status checks: as mentioned in #273, what do people think about requiring an approved review before merging?

My opinion: For maintainers, it doesn't matter as they can usually override that setting and merge anyway (unless you want to disable that). For external contributors, they can't merge anyway as they don't have write permission to the main repo. So it only really makes a difference if you have people with write access to this repo, who are not maintainers. Then you can use this to avoid them to accidentally merging something without a review from a maintainer.

**Comment by yousefmoazzam** [comment]
> > Question for other people's opinion relating to status checks: as mentioned in #273, what do people think about requiring an approved review before merging?
> 
> My opinion: For maintainers, it doesn't matter as they can usually override that setting and merge anyway (unless you want to disable that). For external contributors, they can't merge anyway as they don't have write permission to the main repo. **_So it only really makes a difference if you have people with write access to this repo, who are not maintainers._** Then you can use this to avoid them to accidentally merging something without a review from a maintainer.

That's a useful nuance to be aware of, thanks! For our case, I think we're unlikely to have a contributor with write access to the repo but isn't a maintainer, so on those grounds, it doesn't make much sense to force an approved review, I'd agree with that.

I think a lot of the reason why I'd like the forced approved review is not so much for technical reasons, but more for work culture reasons. I like the idea of encouraging devs to look at the code changes and do code reviews, and not relying solely on the status checks that are in some sense summarising the work that's been done through either a tick or a cross. I'm certainly guilty of this at times too, to not bother reading the code changes and to only look at the status checks, but I'd also say it's more fruitful (and frankly, more fun), to take a look! :slightly_smiling_face: 

But on the other hand, I do understand that people can be busy and not have time to do a code review, even 10/15 mins for a quick look through can be hard to get done - in which case, forcing an approved review is just forcing a few extra clicks in the github GUI.

Maybe that particular mechanism for checking PR readiness should be considered at a later date, and to leave it out of this PR for now.

**Comment by dkazanc** [comment]
I'd say that it is not necessary really, I agree with the @team-gpu comment above. There is unfortunately not enough time to  review every single PR and there are not a lot of reviewers as well!

**Comment by dkazanc** [comment]
I'll merge it now with the FBP memoryhook failing tests that will be fixed later. And I believe Diad and 360 pipelines will pass when we update httomolibgpu and recent dezinger fixes that are probably not in the 2.0 version that gets installed.

**Comment by dkazanc** [comment]
my black was a bit outdated apparently when I tried to fix it. I've fixed it in `main` right after merge

**Comment by yousefmoazzam** [comment]
Ok fair enough. I didn't see another PR that fixed it; did I miss it, or are we still allowing pushing to `main`?

**Pull request merged**